Letterbook.Web/Pages/Profile.cshtml

```diff
@@ -31,15 +31,15 @@
         <h3>Info</h3>
         <div>
             <div>
-                @Model.GetFollowerCount() Followers
+                @await Model.FollowerCount Followers
```

Comment:
Re: 2. Not quite. The `@` is essentially an escape to run C# instead of HTML. `await` is a C# keyword that ties into the dotnet threading model to run asynchronous code. You can await `Task` objects. The semantics are basically identical to awaiting javascript promises. Behind the scenes, it suspends[^1] the executing thread and yields control back to the task pool. The task pool scheduler will resume the thread after the awaited task completes.

Which gets back to the rest of the question. Yes, it is _possible_ to do that awaiting in the model class. But, we would have to block the thread until all of the async values used in the template are available. Or, we could go back to explicitly awaiting and then setting lots of values in the Get handler so that all of the tasks have already completed before we start rendering the template (which happens some time after the handler returns the `Page()` result;

But! that's also only part of the reason that the value didn't populate before. The query that loads the Profile model doesn't include the following/follower collections, because they're potentially enormous. So, the object we get here just has default empty enumerables behind those properties. And I think that's on me. I should have thought about that back when you were building this out originally.

There's a couple of patterns that I've tried to set up to work around that.

1. Query eagerly for specific relations. I expect it will be very common to want to know whether two profiles follow each other, and so we have established queries for that.
2. This is pretty new, but you can build queries as needed for the associated records of a model object you already have. This is how I'm getting the follow/follower counts. If you check the `PageModel` class, you can see that property is defined as `Task<int> FollowerCount => _profiles.GetFollowerCount(_profile);` Which brings us back to why it needs to be awaited.

[^1]: Usually. But it's possible the task has already completed and the method will continue synchronously.

Reply:
So, I would consider putting property setting logic back into the Get handler, if you think it's worth doing. I mainly thought it was likely to get repetitive and error prone. And awaiting things is a really normal thing to do in dotnet. The seamlessness of multithreaded asynchronous programming in C# is a major feature of the language, await is a big part of that.